internal/topology/upgrade/clusterctl_upgrade_test.go

```diff
@@ -0,0 +1,1176 @@
+	g.Expect(err).To(Succeed())
+	refObj, err = getReferencedObject(ctx, env.GetClient(), cpInfraRef, "v1beta2", cluster.Namespace)
+	g.Expect(err).To(Succeed())
+	controlPlaneInfraRef := clusterv1.ContractVersionedObjectReference{
+		APIGroup: refObj.GroupVersionKind().Group,
+		Kind:     refObj.GetKind(),
+		Name:     refObj.GetName(),
+	}
+	g.Expect(refsAfter[controlPlaneInfraRef]).To(Equal(refsBefore[controlPlaneInfraRef]), "controlPlane.spec.machineTemplate.infrastructureRef unexpected change triggering a rollout")
+
+	machineDeployments := &clusterv1.MachineDeploymentList{}
+	err = env.List(ctx, machineDeployments, client.InNamespace(cluster.Namespace), client.MatchingLabels{
+		clusterv1.ClusterNameLabel:          cluster.Name,
+		clusterv1.ClusterTopologyOwnedLabel: "",
+	})
+	g.Expect(err).To(Succeed())
+	g.Expect(machineDeployments.Items).To(HaveLen(1))
+
+	for _, md := range machineDeployments.Items {
+		mdRef := clusterv1.ContractVersionedObjectReference{
+			APIGroup: clusterv1.GroupVersion.Group,
+			Kind:     "MachineDeployment",
+			Name:     md.GetName(),
+		}
+		g.Expect(refsAfter[mdRef]).To(Equal(refsBefore[mdRef]), "machineDeployment "+md.Name+" unexpected change")
+
+		refObj, err = getReferencedObject(ctx, env.GetClient(), &md.Spec.Template.Spec.InfrastructureRef, "v1beta2", cluster.Namespace)
+		g.Expect(err).To(Succeed())
+		mdInfraRef := clusterv1.ContractVersionedObjectReference{
+			APIGroup: refObj.GroupVersionKind().Group,
+			Kind:     refObj.GetKind(),
+			Name:     refObj.GetName(),
+		}
+		g.Expect(refsAfter[mdInfraRef]).To(Equal(refsBefore[mdInfraRef]), "machineDeployment "+md.Name+" .spec.template.spec.infrastructureRef unexpected change triggering a rollout")
+
+		refObj, err = getReferencedObject(ctx, env.GetClient(), md.Spec.Template.Spec.Bootstrap.ConfigRef, "v1beta2", cluster.Namespace)
+		g.Expect(err).To(Succeed())
+		mdBootstrapRef := clusterv1.ContractVersionedObjectReference{
+			APIGroup: refObj.GroupVersionKind().Group,
+			Kind:     refObj.GetKind(),
+			Name:     refObj.GetName(),
+		}
+		g.Expect(refsAfter[mdBootstrapRef]).To(Equal(refsBefore[mdBootstrapRef]), "machineDeployment "+md.Name+" .spec.template.spec.bootstrap.configRef unexpected change triggering a rollout")
+	}
+}
+
+// assertClusterTopologyBecomesStable checks a cluster topology becomes stable ensuring all the objects included cluster, md and referenced template or referencedObjects do not changed/increased generation.
+func assertClusterTopologyBecomesStable(g *WithT, refs map[clusterv1.ContractVersionedObjectReference]int64, namespace, version string) {
+	g.Consistently(func(g Gomega) {
+		for r, generation := range refs {
+			obj := &unstructured.Unstructured{}
+			obj.SetGroupVersionKind(r.GroupKind().WithVersion(version))
+			if r.Kind == "Cluster" || r.Kind == "MachineDeployment" {
+				obj.SetGroupVersionKind(clusterv1.GroupVersion.WithKind(r.Kind))
+			}
```

Comment:
I think this should not be necessary anymore once the last one here is fixed: https://github.com/kubernetes-sigs/cluster-api/pull/12404#discussion_r2168938717